### PR TITLE
ucp: add APIs for request-based put and get operations

### LINKED ISSUE #2320: Add support for get/put with callback
I have been trying to verify the code in osc/ucx in Open MPI and I find it really puzzling that a fence and atomic are needed to get a callback on a particular get/put operation. Is there a technical or performance reason to not have callback versions of the put and get calls for these cases?

Also, in my testing on Aries I find that fence + atomic does not actually ensure completion of all proceeding RDMA operations. That is a separate issue but would not have been discovered if not for this odd (IMHO) API choice.

### PR DESCRIPTION
**Pull request by hjelmn**
For performant MPI request-based RMA it is necessary to be able to
track the completion of individual RMA operations. The current way
this is supported is by calling ucp_worker_fence() followed by issuing
a request-based AMO against the target. This is inefficient. To
address the problem this commit adds two new API calls: ucp_put_nb(),
and ucp_get_nb(). These calls take a user-specified callback that will
be executed when the operation is complete (if the operation does not
complete immediately).

Fixes #2320

This is still a work in progress. It needs feedback and further testing.

Signed-off-by: Nathan Hjelm <hjelmn@lanl.gov>

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6322/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6323/ for details (Mellanox internal link).

**Comment by brminich** [comment]
:+1:  looks ok to me,
@yosefe, @evgeny-leksikov, should not this be aligned with new send/recv API, which is WIP?

**Comment by evgeny-leksikov** [comment]
@brminich new API is not ready yet to align something on that. These have send completion semantic, nothing new from quick view... We will port it as well if new API is accepted.

File: src/ucp/rma/basic_rma.c [hunk 1]
```diff
@@ -377,6 +426,32 @@ ucs_status_t ucp_put_nbi(ucp_ep_h ep, const void *buffer, size_t length,
...
+    UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
+
+    status = UCP_RKEY_RESOLVE(rkey, ep, rma);
+    if (status != UCS_OK) {
+        ptr_status = UCS_STATUS_PTR(status);
+        goto out_unlock;
+    }
+
+    rma_config = &ucp_ep_config(ep)->rma[rkey->cache.rma_lane];
+    ptr_status = ucp_rma_nonblocking_cb(ep, buffer, length, remote_addr, rkey,
```
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: for short puts, would be more optimal to try sending them immediately before allocating a request (e.g https://github.com/openucx/ucx/blob/master/src/ucp/tag/tag_send.c#L177)

File: src/ucp/rma/basic_rma.c [hunk 2]
```diff
@@ -399,6 +474,32 @@ ucs_status_t ucp_get_nbi(ucp_ep_h ep, void *buffer, size_t length,
     return status;
 }
 
+ucs_status_ptr_t ucp_get_nb(ucp_ep_h ep, void *buffer, size_t length,
+                        uint64_t remote_addr, ucp_rkey_h rkey,
```
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: indentation

**Comment by yosefe** [comment]
@shamisp @bbenton pls review as well

File: src/ucp/rma/basic_rma.c (hunk 1 above)
Comment (hjelmn) [reply src/ucp/rma/basic_rma.c]: Are short puts guaranteed to complete on the target before returning?

Comment (yosefe) [reply src/ucp/rma/basic_rma.c]: no, it only guarantees local completion (send buffer is reusable). However, this is the same level of guarantee that request-based completion provides (when request is completed)

Comment (hjelmn) [reply src/ucp/rma/basic_rma.c]: ah, makes sense. i would at some point like to add a worker option for remote completion.

Comment (yosefe) [reply src/ucp/rma/basic_rma.c]: do we need to know about remote completion per-operation (today we have only global one, with flush)? this already came up few times in the past..

**Comment by hjelmn** [comment]
@yosefe Addressed your comments.

File: src/ucp/rma/basic_rma.c (hunk 1 above)
Comment (hjelmn) [reply src/ucp/rma/basic_rma.c]: Its useful to users of the library. I would think it would be easy to implement as a per-worker feature,

Comment (hjelmn) [reply src/ucp/rma/basic_rma.c]: Or even as a different API call.

Comment (hjelmn) [reply src/ucp/rma/basic_rma.c]: Since it only affects put.

**Comment by yosefe** [comment]
we need also @shamisp and @bbenton approval for API changes

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6350/ for details (Mellanox internal link).

**Comment by hjelmn** [comment]
Probably a failure with a previous push. Had a typo at one point.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6351/ for details (Mellanox internal link).

File: src/ucp/api/ucp.h [hunk 3]
```diff
@@ -2502,6 +2503,53 @@ ucs_status_t ucp_put(ucp_ep_h ep, const void *buffer, size_t length,
...
+ * @return otherwise            - Operation was scheduled and can be
+ *                              completed at any point in time. The request handle
+ *                              is returned to the application in order to track
+ *                              progress of the operation. The application is
+ *                              responsible for releasing the handle using
+ *                              @ref ucp_request_free "ucp_request_free()" routine.
+ */
+ucs_status_ptr_t ucp_put_nb(ucp_ep_h ep, const void *buffer, size_t length,
+                            uint64_t remote_addr, ucp_rkey_h rkey,
+                            ucp_send_callback_t cb);
```
Comment (shamisp) [inline src/ucp/api/ucp.h]: is ucp_send_callback_t is still a good name for the callback if we are using it for put/get ?

File: src/ucp/api/ucp.h [hunk 4]
```diff
@@ -2502,6 +2503,53 @@ ucs_status_t ucp_put(ucp_ep_h ep, const void *buffer, size_t length,
 ucs_status_t ucp_put_nbi(ucp_ep_h ep, const void *buffer, size_t length,
                          uint64_t remote_addr, ucp_rkey_h rkey);
 
+/**
+ * @ingroup UCP_COMM
+ * @brief Non-blocking implicit remote memory put operation.
```
Comment (bbenton) [inline src/ucp/api/ucp.h]: This is a non-blocking put operation, not a non-blocking implicit operation, right?

File: src/ucp/api/ucp.h [hunk 5]
```diff
@@ -2502,6 +2503,53 @@ ucs_status_t ucp_put(ucp_ep_h ep, const void *buffer, size_t length,
...
+ *                           completion.
+ *
+ * @return UCS_OK               - The operation was completed immediately.
+ * @return UCS_PTR_IS_ERR(_ptr) - The operation failed.
+ * @return otherwise            - Operation was scheduled and can be
+ *                              completed at any point in time. The request handle
+ *                              is returned to the application in order to track
+ *                              progress of the operation. The application is
+ *                              responsible for releasing the handle using
+ *                              @ref ucp_request_free "ucp_request_free()" routine.
```
Comment (bbenton) [inline src/ucp/api/ucp.h]: Is a request handle possibly returned? I thought it was just UCS_OK, UCS_INPROGRESS, or an error.

Comment (hjelmn) [reply src/ucp/api/ucp.h]: This is the same behavior as non-blocking amo fetch. The request needs to be returned so the caller can fill out the desired fields of the request. I find this strange but it is consistent with elsewhere in ucx.

File: src/ucp/api/ucp.h (hunk 4 above)
Comment (hjelmn) [reply src/ucp/api/ucp.h]: No idea. C&P'ed text. What is the definition of implicit in UCX?

Comment (shamisp) [reply src/ucp/api/ucp.h]: Implicit means - no request/cb is returned. Good catch. it is non-blocking operation.

Comment (hjelmn) [reply src/ucp/api/ucp.h]: Ah, ok. Will fix the wording.

**Comment by hjelmn** [comment]
Removed the word implicit from the descriptions of the new functions.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6405/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:bgate:retest

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6439/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6455/ for details (Mellanox internal link).

**Comment by hjelmn** [comment]
Hmm, what is going on. This was passing all the tests now I can't even see the results for the non-internal testing.

**Comment by yosefe** [comment]
issues with test setup

**Comment by yosefe** [comment]
bot:bgate:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6458/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@bbenton is it ok now?

**Comment by yosefe** [comment]
@gmegan can you pls take a look?

**Pull request merged**